src/Utility/Instrumentation/InstrumentationScope.cs

```diff
@@ -0,0 +1,219 @@
+﻿using OpenAI.Chat;
+using System;
+using System.Buffers;
+using System.ClientModel;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+
+namespace OpenAI.Instrumentation;
+
+internal class InstrumentationScope : IDisposable
+{
+    private static readonly ActivitySource s_chatSource = new ActivitySource("OpenAI.ChatClient");
+    private static readonly Meter s_chatMeter = new Meter("OpenAI.ChatClient");
+
+    // TODO: add explicit histogram buckets once System.Diagnostics.DiagnosticSource 9.0 is used
+    private static readonly Histogram<double> s_duration = s_chatMeter.CreateHistogram<double>(Constants.GenAiClientOperationDurationMetricName, "s", "Measures GenAI operation duration.");
+    private static readonly Histogram<long> s_tokens = s_chatMeter.CreateHistogram<long>(Constants.GenAiClientTokenUsageMetricName, "{token}", "Measures the number of input and output token used.");
+
+    private readonly string _operationName;
+    private readonly string _serverAddress;
+    private readonly int _serverPort;
+    private readonly string _requestModel;
+
+    private Stopwatch _duration;
+    private Activity _activity;
+    private TagList _commonTags;
+
+    private InstrumentationScope(
+        string model, string operationName,
+        string serverAddress, int serverPort)
+    {
+        _requestModel = model;
+        _operationName = operationName;
+        _serverAddress = serverAddress;
+        _serverPort = serverPort;
+    }
+
+    public static InstrumentationScope StartChat(string model, string operationName,
+        string serverAddress, int serverPort, ChatCompletionOptions options)
+    {
+        if (s_chatSource.HasListeners() || s_tokens.Enabled || s_duration.Enabled)
```

Comment:
Should this check be refactored into a separate routine such that it can be reused in other methods.